### DISCUSSION TITLE
CIP33 – File Storage in P2WSH Outputs

### DISCUSSION THREAD

**Discussion by Jpja** [CounterpartyXCP/CIPs#126] [CIPs]
Use this tread to discuss [CIP33 – File Storage in P2WSH Outputs](https://github.com/CounterpartyXCP/cips/blob/master/cip-0033.md).

**Comment by Jpja** [comment]
**Minting from Electrum**

Use the "XCP Opreturn Builder for Electrum"
- [Token Issuance with File](https://jpja.github.io/Electrum-Counterparty/cip33_issuance.html)
- [Broadcast File](https://jpja.github.io/Electrum-Counterparty/cip33_broadcast.html)

_Network_
Try it out with testnet first. Get coins from a faucet:
https://bitcoinfaucet.uo1.net/
https://coinfaucet.eu/en/btc-testnet/

_UTXO (Input Coin)_
- Under the `Coins` tab select an `Output point` with an amount of at least 0.001.
- Right click -> `Spend`
- Again with the same coin, Right clock -> `Copy` -> `Output point`
![image](https://github.com/CounterpartyXCP/cips/assets/8447505/8c934903-918a-4b4f-8228-df06cc0f333f)

_Token Name_
It is recommended to leave the default random name. If you change it, make sure it has not been used before.

_Token Supply_
The supply will be locked at issuance.

_File_
- Any file is valid but only txt, gif, png, and jpeg will show in the decoder.
- Max size 6 kB
- Use an appropriate filename. The name will also go onchain and be shown in explorers.

Copy the generated transaction to Electrum's `Pay to` field and click `Pay...`.
![image](https://github.com/CounterpartyXCP/cips/assets/8447505/30a3f766-08ec-4f66-a880-efe4685bef2a)
 
Double-check that the input coin is as expected. 
![image](https://github.com/CounterpartyXCP/cips/assets/8447505/276ce516-fcaf-492b-9e3c-6415968615c5)

Once the transaction is confirmed, you can view it any CIP33-compatible Counterparty explorer.

At the moment, only the Counterparty decoder will show the image:
- mainnet: https://jpja.github.io/Electrum-Counterparty/decode_tx?tx=549a5cc4bc189c800f0f9ea01068e8a7fd987c7dadb40c0b6a224d489ed070cc
- testnet:  https://jpja.github.io/Electrum-Counterparty/decode_tx?tx=83fc87b4fffd78c8ffd55469206f0f90c0dbbb620b32e8d4adfb7d46148a07f5&network=testnet
(replace the hash with your Bitcoin tx id)

Explorers without CIP33 support will show the asset and the filename but not the file itself.

**Comment by jotapea** [comment]
> 3. does not require a consensus change (unlike CIP31).

THIS!!! Love it.

What are the drawbacks? Can the dust be redeemed?

**Reply by Jpja** [reply]
P2WSH outputs are burn addresses. Dust cannot be redeemed. This by design since Stamps intend to inscribe into permanent UTXOs. The way they do it now in 2-3 redeemable multisig is not very elegant – and more wasteful than it needs to be.

**Reply by jotapea** [reply]
Ok maybe this drawback should be made more explicit, and in comparison to counterparty-lib's `multisig` encoding, instead of being compared mainly to stamps. The stamps project does promote "unprunable utxo" art, but this is optional by what they call "keyburn". Whoever decides to do this is their freedom of expression, but this is not "responsible" bitcoin usage, an important fact we should not ignore.

I would explain these more clearly, to show that the Counterparty project does care about the long term health of Bitcoin. Eventually, the redeeming of `multisig` outputs can be added to counterparty-lib...

**Reply by Jpja** [reply]
Agree, the burn part is important. I changed the abstract from "encode a file in P2WSH outputs" to "encode a file in burn-addresses".

**Reply by Jpja** [reply]
This CIP was btw in part a response to the default multisig dust being reduced (which I was against).  The default was made so low that it's economically infeasible to redeem. Since people essentially use burn addresses anyway, why not use a method that is 50% more efficient?

**Reply by jotapea** [reply]
Yes, I know that happened. Thankfully it is only a configuration, which can be changed again to a more balanced default.

**Comment by keyuno** [comment]
Excellent work! I haven't had a chance to review all the details thoroughly yet, but at first glance, it appears to be an improvement over the STAMPS project. Additionally, I must admit that, in general, I prefer the more elegant approach of broadcasting IPFS content identifiers as representations of NFTs.

> FILE: - Case insensitive prefix of asset description or broadcast signaling that a file is encoded in the transaction in accordance with CIP33 specifications..

I have a minor suggestion. There appears to be a common confusion between NFTs and representations in the general public. To address this, consider replacing 'FILE' with 'REPRESENTATION' to underscore that NFTs are not representations; individuals acquire and collect tokens, not the representations of those tokens.

**Reply by Jpja** [reply]
Good points! The scope of this CIP is to propose a better way of storing files in **permanent UTXOs**, aka "stamping". Whether or not this makes sense versus redeemable outputs, or hash references, etc is indeed up for debate. In the end I believe the answer boils down to preferences. Since stamps currently use 2-3 redeemable multisig and intend never to sweep the dust, I believe CIP33 provides a more elegant approach to their concept.

**Reply by keyuno** [reply]
> The scope of this CIP is to propose a better way of storing files in permanent UTXOs, aka "stamping". Whether or not this makes sense versus redeemable outputs, or hash references, etc is indeed up for debate.

I would like to add that the discussion on permanence is non-technical and often leads to confusion about the true nature of provably-prunable outputs. To compromise the persistence of provably-prunable outputs, one would need to disrupt the integrity of Bitcoin. Therefore, I wouldn't characterize it as a 'debate,' as the notion of dividing assets based on permanence is fundamentally not a viable starting point.

**Reply by jotapea** [reply]
Agree. We using Bitcoin as a medium, will always need a full-node that indexes all transactions anyway. The branding slogans are more about the meme... and it works! So there is little we can debate or fight, but we can't ignore.

**Comment by jdogresorg** [comment]
NACK

Overall I agree with the thought process behind moving files out of the CP database. I have some fundamental concerns tho that  prevent me from supporting this CIP.

1.  CP features should be encoding agnostic (all features should support multiple encoding methods), not just force one form of encoding (and an inefficient/expensive method at that). What about when CP gets P2WSH/taproot support, we gonna write a separate CIP for every encoding method? 

2. Forced keyburn is a non-starter for me. People should be free to make their own decision on if they want to redeem their dust, not be forced into very destructive / irresponsible behavior. That type of behavior might be fine for an individual project, but we gotta hold ourselves to a higher standard on the protocol level.

3. Multisig encoding is very abusive to the UTXO set and should be avoided unless no other options are available . We should be focusing on taproot / p2wsh encoding and storing more data for cheaper in prunable witness areas.

CIP31 is encoding agnostic, has a PR with functioning code, and does not store files in the database.
https://github.com/CounterpartyXCP/cips/blob/master/cip-0031.md

**Reply by Jpja** [reply]
I noticed a mistake in the proposal. I listed the CIP as `Type: Standards`. Just changed this to `Informational` as this CIP describes a method of using Bitcoin/Counterparty without changes to the protocol or client. 

With this out of the way, a rejection can only be for technical reasons, not judgmental ones (ref CIP 1). Regarding your comments:

1. The file is encoded in P2WSH outputs outside of Counterparty consensus. I considered allowing P2TR outputs too as these have the same properties (32 bytes and 330 dust minimum). I chose P2WSH only for simplicity.
2. Burn became the de-facto default for Counterparty after the latest upgrade – despite my objections. The dust amount for multisig encoding was made so low that it became uneconomical to redeem. This CIP is partly a response to this. Since people are burning dust anyway, at least do it in a less harmful way (<50% blockchain space used, no impact on the Counterparty DB).
3. That would be a different approach, worthy of a separate CIP.

CIP31 is a change of the protocol rules. The code implementation truncates descriptions, which is a huge deal. CIP33 is an optional method with functional encoder and decoder and which, if widely used, will decrease the burden on node operators. I truly think it deserves to become a formal CIP.

**Reply by jdogresorg** [reply]
> Just changed this to Informational as this CIP

Thank you for the clarification. As this is no longer a proposal to change `counterparty-lib`, my objections to this CIP are removed... continue the experimentation.

> Burn became the de-facto default for Counterparty after the latest upgrade...

Multisig dust was simply reduced from 7800 sats to 1000 sats, that is all that changed... no "burn". 

I feel it is important to keep terminology clear.  "Burn" is destroying something in a provably unspendable way. The multisig outputs CP txs generate are still spendable, and always have been.

> The dust amount for multisig encoding was made so low that it became uneconomical to redeem.

"Uneconomical to redeem" is a matter of opinion... we can agree to disagree :)

Just over 2 months ago I gathered over 40,000 UTXOS (1000 sats each)  paying only 1.63 sats/byte. 

https://mempool.space/tx/46604b88deeefee66a60c0a0e161f6adbfec037f1793c3ab9d0559abee7b8bdb 
https://mempool.space/tx/b56b6ff1768de9bd6ee62e0ceb06b3e8fabb178dcf2d90260a5bd3aca7077c26 
https://mempool.space/tx/7b4111f57b2947ab89605705725c87711d37c41926489b5766e7962b2a510417
https://mempool.space/tx/9e8336a345086a0943a698d0c5d858db218a5d40a5f0f6e006f4df77a294970c 
https://mempool.space/tx/dcdbb5570ff899815386b686dbeece3cd60b5eafea26912fbfc6c924ff004abc   
... etc

Happy to see you playing around n becoming more involved in dev 👍

**Reply by jotapea** [reply]
> The code implementation truncates descriptions, which is a huge deal.

Agree. This is my main issue with CIP31. Very inelegant and does nothing to solve the unnecessary `description` copying, which is a huge part of the SQLite DB bloat problem.
